Add cloneable flag to ShadowRootInit

rniwa opened the pull request:
#### d7d9268e700a19cd1b06de3920229b83f9071ad9
<pre>
Add cloneable flag to ShadowRootInit
<a href="https://bugs.webkit.org/show_bug.cgi?id=249369">https://bugs.webkit.org/show_bug.cgi?id=249369</a>

Reviewed by Chris Dumez.

Add the support for cloneable boolean flag on ShadowRootInit. ShadowRoot and its descendant nodes becomes cloneable
when this flag is set. The flag is set by default for declarative shadow DOM.

See <a href="https://github.com/whatwg/dom/issues/1137">https://github.com/whatwg/dom/issues/1137</a> for discussion.

* LayoutTests/fast/shadow-dom/cloneable-shadow-root-expected.txt: Added.
* LayoutTests/fast/shadow-dom/cloneable-shadow-root.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/declarative-shadow-dom-basic.tentative-expected.txt:
* LayoutTests/platform/win/TestExpectations

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::cloneNodeInternal):
(WebCore::Element::cloneShadowTreeIfPossible): Added.
(WebCore::Element::cloneElementWithChildren):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ShadowRoot.cpp:s
(WebCore::ShadowRoot::ShadowRoot):
(WebCore::ShadowRoot::childrenChanged):
(WebCore::ShadowRoot::cloneNodeInternal): Added the implementation.
* Source/WebCore/dom/ShadowRoot.h:
(WebCore::ShadowRoot): Renamed m_type to m_mode for consistency &amp; clarity.
* Source/WebCore/dom/ShadowRootInit.h:
* Source/WebCore/dom/ShadowRootInit.idl:
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::HTMLConstructionSite::insertHTMLTemplateElement): Allow declarative shadow DOM inside in a template content.

Canonical link: <a href="https://commits.webkit.org/257978@main">https://commits.webkit.org/257978@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/70ebd8d70f5094118cc623fb0d6d1c23df91ab6b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100546 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9687 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33587 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109850 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170140 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104531 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10621 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92941 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107705 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106325 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/8033 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/91281 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34645 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89945 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22670 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77614 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3411 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24188 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3419 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9532 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43679 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5217 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2856 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -560,18 +560,37 @@ Ref<Node> Element::cloneNodeInternal(Document& targetDocument, CloningOperation
+    if (!oldShadowRoot || !oldShadowRoot->isCloneable())
+        return;
+
+    Ref clone = static_cast<Node&>(*oldShadowRoot).cloneNodeInternal(newHost.document(), Node::CloningOperation::SelfWithTemplateContent);
```
We should make cloneNodeInternal() public in ShadowRoot instead of doing this cast. There is no good reason for this function to be public on Node and private on ShadowRoot AFAICT. Also, it prevents de-virtualization of the function call.

rniwa replied:
That's a good point. Will do.

cdumez commented on Source/WebCore/dom/ShadowRoot.cpp:
```diff
@@ -225,10 +225,20 @@ void ShadowRoot::setResetStyleInheritance(bool value)
+    }
+
+    RELEASE_ASSERT_NOT_REACHED(); // ShadowRoot is never cloned directly on its own.
+    return static_cast<Node&>(*this);
```
I am a little surprised that the static_cast is needed.

rniwa replied:
That's definitely not necessary. Removed.

cdumez commented on Source/WebCore/dom/ShadowRoot.h:
```diff
@@ -48,13 +48,14 @@ class ShadowRoot final : public DocumentFragment, public TreeScope {
 
-    enum class DelegatesFocus : uint8_t { Yes, No };
-    enum class AvailableToElementInternals : uint8_t { Yes, No };
+    enum class DelegatesFocus : uint8_t { No, Yes };
```
Could use `: bool`

rniwa replied:
Okay, changed.

cdumez commented on Source/WebCore/dom/ShadowRoot.h:
```diff
@@ -48,13 +48,14 @@ class ShadowRoot final : public DocumentFragment, public TreeScope {
-    enum class DelegatesFocus : uint8_t { Yes, No };
-    enum class AvailableToElementInternals : uint8_t { Yes, No };
+    enum class DelegatesFocus : uint8_t { No, Yes };
+    enum class Cloneable : uint8_t { No, Yes };
```
ditto.

cdumez commented on Source/WebCore/dom/ShadowRoot.h:
```diff
@@ -48,13 +48,14 @@ class ShadowRoot final : public DocumentFragment, public TreeScope {
-    enum class AvailableToElementInternals : uint8_t { Yes, No };
+    enum class DelegatesFocus : uint8_t { No, Yes };
+    enum class Cloneable : uint8_t { No, Yes };
+    enum class AvailableToElementInternals : uint8_t { No, Yes };
```
ditto.

cdumez commented on Source/WebCore/dom/ShadowRoot.h:
```diff
@@ -77,6 +78,8 @@ class ShadowRoot final : public DocumentFragment, public TreeScope {
     bool containsFocusedElement() const { return m_containsFocusedElement; }
     void setContainsFocusedElement(bool flag) { m_containsFocusedElement = flag; }
 
+    bool isCloneable() const { return m_cloneable; }
```
Could be nice to name the data member `m_isCloneable` too as per the style for boolean variables (needs a prefix).

rniwa replied:
Fixed.

cdumez reviewed (approved):
LGTM

rniwa committed:
Add cloneable flag to ShadowRootInit
https://bugs.webkit.org/show_bug.cgi?id=249369

Reviewed by Chris Dumez.

Add the support for cloneable boolean flag on ShadowRootInit. ShadowRoot and its descendant nodes becomes cloneable
when this flag is set. The flag is set by default for declarative shadow DOM.

See https://github.com/whatwg/dom/issues/1137 for discussion.

* LayoutTests/fast/shadow-dom/cloneable-shadow-root-expected.txt: Added.
* LayoutTests/fast/shadow-dom/cloneable-shadow-root.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/declarative/declarative-shadow-dom-basic.tentative-expected.txt:
* LayoutTests/platform/win/TestExpectations

* Source/WebCore/dom/Element.cpp:
(WebCore::Element::cloneNodeInternal):
(WebCore::Element::cloneShadowTreeIfPossible): Added.
(WebCore::Element::cloneElementWithChildren):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ShadowRoot.cpp:s
(WebCore::ShadowRoot::ShadowRoot):
(WebCore::ShadowRoot::childrenChanged):
(WebCore::ShadowRoot::cloneNodeInternal): Added the implementation.
* Source/WebCore/dom/ShadowRoot.h:
(WebCore::ShadowRoot): Renamed m_type to m_mode for consistency & clarity.
* Source/WebCore/dom/ShadowRootInit.h:
* Source/WebCore/dom/ShadowRootInit.idl:
* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::HTMLConstructionSite::insertHTMLTemplateElement): Allow declarative shadow DOM inside in a template content.

Canonical link: https://commits.webkit.org/257978@main
